gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,432 @@
+    ind: int = 0,
+) -> Tensor:
+    """
+    A helper file for topological regularizer
+    Given a neural network and a dataset, computes the
+    ephemeral critical values and points of the linear
+    Morse function defined by the dataset and connections.
+    The filtering directions is specified by ind (0: Positive, 1 negative)
+    A helper file for topological regularizer
+    Args:
+        model: torch model
+        coords: torch.tensor, the nodes of the linear surrogate function
+        dummy_loader: torch dataloader containing the coords.
+            connections: numpy_array. An n by 2 numpy array describing
+            pairwise connections between the nodes.
+        expansion: The size of the clique expansion
+        ind: integer (for now boolean) describing filtration direction
+            0: positive filtration, 1: negative filtration
+    Returns
+    -------
+        coordinates: The coordinates of the critical points that
+        contribute to the complexity of the decision boundary.
+    """
+    res = evaluate_model_on_grid(dummy_loader, net)
+    if ind == 1:
+        res = 1 - res
+    out_pers, persts = get_persistent_pairs(res, connections, expansion)
+
+    """ Get only the persistent pairs that contribute to
+    the complexity of the decision boundary:
+    These are the ones that are on the different side of the boundary
+    """
+    tmp = out_pers - 0.5
+    cuts = (tmp[:, 0] * tmp[:, 1]) < 0
+
+    goodinds = np.argmin(abs(tmp[cuts, :]), 1)
+    tmp_inds = persts[cuts, :]
+    indices = np.zeros(tmp_inds.shape[0])
+    for i in range(tmp_inds.shape[0]):
+        sele = goodinds[i]
+        indices[i] = tmp_inds[i, sele]
+    # The good values
+    indices = indices.astype(int)
+    coordinates = coords[indices, :].clone().detach().requires_grad_(True)
+    return coordinates
+
+
+class Regularizer:
+    def __init__(self, **kwargs):
+        self.regularizer_arguments = kwargs
+
+    """
+    An abstract class for handling various regularization schemes.
+    A Regularizer needs to have 3 methods:
+    - preprocess
+    - update_params
+    - regularization_penalty
+    """
+
+    @abstractmethod
+    def preprocess(self, **regularizer_arguments: Any) -> Any:
+        """
+        performs preprocessing for the regularizer, if any is necessary
+
+        """
+        return regularizer_arguments
+
+    @abstractmethod
+    def update_params(self, pre_processed_arguments: Any) -> Any:
+        """
+        Any processing needed in a forward pass of the network
+        Anything that doesn't require gradients can go here
+        """
+        return pre_processed_arguments
+
+    @abstractmethod
+    def regularization_penalty(
+        self, forward_passed_arguments: Any, model: torch.nn.Module
+    ) -> Tensor:
+        """
+        Computes the actual regression penalty.
+        This must be an auto-differentiable function of its inputs
+        """
+        return torch.sum(forward_passed_arguments) ** 2
```

Comment:
This method is abstract, so replace implementation by `pass`

Reply:
done